src/uu/tail/src/tail.rs

```diff
@@ -125,127 +187,302 @@ impl Settings {
+                    } else {
+                        reader = BufReader::new(file);
+                        unbounded_tail(&mut reader, &settings)?;
+                    }
+                    if settings.follow.is_some() {
+                        // Insert existing/file `path` into `files.map`
+                        files.insert(
+                            path.canonicalize()?,
+                            PathData {
+                                reader: Some(Box::new(reader)),
+                                metadata,
+                                display_name,
+                            },
+                        );
+                    }
+                }
+                Err(e) if e.kind() == std::io::ErrorKind::PermissionDenied => {
+                    settings.exit_code = 1;
+                    show_error!(
+                        "cannot open {} for reading: Permission denied",
+                        display_name.quote()
+                    );
+                }
+                Err(e) => {
+                    return Err(USimpleError::new(
+                        1,
+                        format!("{}: {}", display_name.quote(), e),
+                    ));
+                }
+            }
+        } else if settings.retry && settings.follow.is_some() {
+            if path.is_relative() {
+                path = std::env::current_dir()?.join(&path);
+            }
+            // Insert non-is_tailable() paths into `files.map`
+            files.insert(
+                path.to_path_buf(),
+                PathData {
+                    reader: None,
+                    metadata,
+                    display_name,
+                },
+            );
         }
     }
 
-    if settings.follow {
-        follow(&mut readers[..], settings)?;
+    if settings.follow.is_some() {
+        /*
+        POSIX specification regarding tail -f
+        If the input file is a regular file or if the file operand specifies a FIFO, do not
+        terminate after the last line of the input file has been copied, but read and copy
+        further bytes from the input file when they become available. If no file operand is
+        specified and standard input is a pipe or FIFO, the -f option shall be ignored. If
+        the input file is not a FIFO, pipe, or regular file, it is unspecified whether or
+        not the -f option shall be ignored.
+        */
+        if files.map.is_empty() || !files.files_remaining() && !settings.retry {
+            if !files.only_stdin_remaining() {
+                show_error!("{}", text::NO_FILES_REMAINING);
+            }
+        } else if !(settings.stdin_is_pipe_or_fifo && settings.paths.len() == 1) {
+            follow(&mut files, &mut settings)?;
+        }
+    }
+
+    if settings.exit_code > 0 {
+        #[cfg(unix)]
+        if stdin_is_bad_fd() {
+            show_error!("-: {}", text::BAD_FD);
+        }
+        return Err(USimpleError::new(settings.exit_code, ""));
```

Comment:
Here again, if `Ok` is returned from `uumain` the exit code will be determined from previous `show!` and `set_exit_code` calls, so this is unnecessary if you use that mechanism.